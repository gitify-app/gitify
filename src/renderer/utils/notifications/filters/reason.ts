import type { AccountNotifications, SettingsState } from '../../../types';
import type { Notification, Reason } from '../../../typesGitHub';

export function hasReasonFilters(settings: SettingsState) {
  return settings.filterReasons.length > 0;
}

export function isReasonFilterSet(settings: SettingsState, reason: Reason) {
  return settings.filterReasons.includes(reason);
}

export function getReasonFilterCount(
  notifications: AccountNotifications[],
  reason: Reason,
) {
  return notifications.reduce(
    (sum, account) =>
      sum + account.notifications.filter((n) => n.reason === reason).length,
    0,
  );
}

export function filterNotificationByReason(
  notification: Notification,
  reason: Reason,
): boolean {
  return notification.reason === reason;
}
