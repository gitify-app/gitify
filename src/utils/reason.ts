import type { FormattedReason } from '../types';
import type { Reason } from '../typesGitHub';

const FORMATTED_REASONS: Record<Reason, FormattedReason> = {
  approval_requested: {
    title: 'Approval Requested',
    description: 'You were requested to review and approve a deployment.',
  },
  assign: {
    title: 'Assigned',
    description: 'You were assigned to the issue.',
  },
  author: {
    title: 'Authored',
    description: 'You created the thread.',
  },
  ci_activity: {
    title: 'Workflow Run Completed',
    description:
      'A GitHub Actions workflow run was triggered for your repository.',
  },
  comment: {
    title: 'Commented',
    description: 'You commented on the thread.',
  },
  invitation: {
    title: 'Invitation Received',
    description: 'You accepted an invitation to contribute to the repository.',
  },
  manual: {
    title: 'Updated',
    description: 'You subscribed to the thread (via an issue or pull request).',
  },
  member_feature_requested: {
    title: 'Member Feature Requested',
    description:
      'Organization members have requested to enable a feature such as Draft Pull Requests or Copilot.',
  },
  mention: {
    title: 'Mentioned',
    description: 'You were specifically @mentioned in the content.',
  },
  review_requested: {
    title: 'Review Requested',
    description:
      "You, or a team you're a member of, were requested to review a pull request.",
  },
  security_advisory_credit: {
    title: 'Security Advisory Credit Received',
    description: 'You were credited for contributing to a security advisory.',
  },
  security_alert: {
    title: 'Security Alert Received',
    description:
      'GitHub discovered a security vulnerability in your repository.',
  },
  state_change: {
    title: 'State Changed',
    description:
      'You changed the thread state (for example, closing an issue or merging a pull request).',
  },
  subscribed: {
    title: 'Updated',
    description: "You're watching the repository.",
  },
  team_mention: {
    title: 'Team Mentioned',
    description: 'You were on a team that was mentioned.',
  },
};

const UNKNOWN_REASON: FormattedReason = {
  title: 'Unknown',
  description: 'The reason for this notification is not supported by the app.',
};

export function formatReason(reason: Reason): FormattedReason {
  return FORMATTED_REASONS[reason] || UNKNOWN_REASON;
}
