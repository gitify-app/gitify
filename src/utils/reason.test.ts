import type { Reason } from '../typesGitHub';
import { formatReason } from './reason';

describe('utils/reason.ts', () => {
  it('formatReason - should format the notification reason', () => {
    expect(formatReason('approval_requested')).toMatchSnapshot();
    expect(formatReason('assign')).toMatchSnapshot();
    expect(formatReason('author')).toMatchSnapshot();
    expect(formatReason('ci_activity')).toMatchSnapshot();
    expect(formatReason('comment')).toMatchSnapshot();
    expect(formatReason('invitation')).toMatchSnapshot();
    expect(formatReason('manual')).toMatchSnapshot();
    expect(formatReason('member_feature_requested')).toMatchSnapshot();
    expect(formatReason('mention')).toMatchSnapshot();
    expect(formatReason('review_requested')).toMatchSnapshot();
    expect(formatReason('security_advisory_credit')).toMatchSnapshot();
    expect(formatReason('security_alert')).toMatchSnapshot();
    expect(formatReason('state_change')).toMatchSnapshot();
    expect(formatReason('subscribed')).toMatchSnapshot();
    expect(formatReason('team_mention')).toMatchSnapshot();
    expect(formatReason('something_else_unknown' as Reason)).toMatchSnapshot();
  });
});
