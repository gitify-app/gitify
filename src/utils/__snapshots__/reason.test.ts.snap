// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/reason.ts formatReason - should format the notification reason 1`] = `
{
  "description": "You were requested to review and approve a deployment.",
  "title": "Approval Requested",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 2`] = `
{
  "description": "You were assigned to the issue.",
  "title": "Assigned",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 3`] = `
{
  "description": "You created the thread.",
  "title": "Authored",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 4`] = `
{
  "description": "A GitHub Actions workflow run was triggered for your repository.",
  "title": "Workflow Run Completed",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 5`] = `
{
  "description": "You commented on the thread.",
  "title": "Commented",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 6`] = `
{
  "description": "You accepted an invitation to contribute to the repository.",
  "title": "Invitation Received",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 7`] = `
{
  "description": "You subscribed to the thread (via an issue or pull request).",
  "title": "Updated",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 8`] = `
{
  "description": "Organization members have requested to enable a feature such as Draft Pull Requests or Copilot.",
  "title": "Member Feature Requested",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 9`] = `
{
  "description": "You were specifically @mentioned in the content.",
  "title": "Mentioned",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 10`] = `
{
  "description": "You, or a team you're a member of, were requested to review a pull request.",
  "title": "Review Requested",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 11`] = `
{
  "description": "You were credited for contributing to a security advisory.",
  "title": "Security Advisory Credit Received",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 12`] = `
{
  "description": "GitHub discovered a security vulnerability in your repository.",
  "title": "Security Alert Received",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 13`] = `
{
  "description": "You changed the thread state (for example, closing an issue or merging a pull request).",
  "title": "State Changed",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 14`] = `
{
  "description": "You're watching the repository.",
  "title": "Updated",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 15`] = `
{
  "description": "You were on a team that was mentioned.",
  "title": "Team Mentioned",
}
`;

exports[`utils/reason.ts formatReason - should format the notification reason 16`] = `
{
  "description": "The reason for this notification is not supported by the app.",
  "title": "Unknown",
}
`;
